proto/lbm/token/v1/event.proto

```diff
@@ -3,75 +3,168 @@ package lbm.token.v1;
-message EventRevoke {
-  // class id associated with the token.
-  string class_id = 1;
+// EventGrant is emitted when a granter grants its permission to a grantee.
+//
+// Since: 0.46.0 (finschia)
+message EventGrant {
+  // contract id associated with the token class.
+  string contract_id = 1;
+  // address of the granter.
+  string granter = 2;
+  // address of the grantee.
+  string grantee = 3;
+  // permission on the token class.
+  string permission = 4;
+}
 
-  // address of the revoked account.
+// EventAbandon is emitted when a grantee abandons its permission.
+//
+// Since: 0.46.0 (finschia)
+message EventAbandon {
+  // contract id associated with the token class.
+  string contract_id = 1;
+  // address of the grantee which abandons its grant.
   string grantee = 2;
-
-  // action on the token class. Must be one of "mint", "burn" and "modify".
-  string action = 3;
+  // permission on the token class.
+  string permission = 3;
 }
 
-// EventMint is emitted on Msg/Mint
-message EventMint {
-  // class id associated with the token.
-  string class_id = 1;
-  string to       = 2;
-  string amount   = 3 [(gogoproto.customtype) = "github.com/line/lbm-sdk/types.Int", (gogoproto.nullable) = false];
+// EventMinted is emitted when tokens are minted.
+//
+// Since: 0.46.0 (finschia)
+message EventMinted {
+  // contract id associated with the token class.
+  string contract_id = 1;
+  // address which triggered the mint.
+  string operator = 2;
+  // recipient of the tokens.
+  string to = 3;
+  // number of tokens minted.
+  string amount = 4 [(gogoproto.customtype) = "github.com/line/lbm-sdk/types.Int", (gogoproto.nullable) = false];
 }
 
-// EventBurn is emitted on Msg/Burn
-message EventBurn {
-  // class id associated with the token.
-  string class_id = 1;
-  string from     = 2;
-  string amount   = 3 [(gogoproto.customtype) = "github.com/line/lbm-sdk/types.Int", (gogoproto.nullable) = false];
+// EventBurned is emitted when tokens are burnt.
+//
+// Since: 0.46.0 (finschia)
+message EventBurned {
+  // contract id associated with the token class.
+  string contract_id = 1;
+  // address which triggered the burn.
+  string operator = 2;
+  // holder whose tokens were burned.
+  string from = 3;
+  // number of tokens burned.
+  string amount = 4 [(gogoproto.customtype) = "github.com/line/lbm-sdk/types.Int", (gogoproto.nullable) = false];
 }
 
-// EventModify is emitted on Msg/Modify
-message EventModify {
-  // class id associated with the token.
-  string class_id = 1;
-  string key      = 2;
-  string value    = 3;
+// EventModified is emitted when the information of a token class is modified.
+//
+// Since: 0.46.0 (finschia)
+message EventModified {
+  // contract id associated with the token class.
+  string contract_id = 1;
+  // address which triggered the modify.
+  string operator = 2;
+  // changes on the metadata of the class.
+  repeated Pair changes = 3 [(gogoproto.nullable) = false];
```

Comment:
How about emitting multiple events rather than one event with multiple changes?

For example, let's think about a case where `name` and `symbol` are modified.

(1) In case of your idea, the following one event will be emitted.

```
{
  "contract_id": "<contract ID>",
  "operator": "<operator>",
  "changes": [
    { "key": "name", "value": "<new name>" },
    { "key": "symbol", "value": "<new symbol>" }
  ]
}
```

(2) In another idea, the following two events will be emitted.

```
{
  "contract_id": "<contract ID>",
  "operator": "<operator>",
  "field": "name",
  "value": "<new name>"
}

{
  "contract_id": "<contract ID>",
  "operator": "<operator>",
  "field": "symbol",
  "value": "<new symbol>"
}
```

**The advantage of (2) is, clients do not need to parse JSON** to interpret modify events.
In case of (1), clients will need to parse JSON because, as you said before, the field `changes` will be a JSON string in transactions which clients receive via Mainnet RPC.

Reply:
> **The advantage of (2) is, clients do not need to parse JSON** to interpret modify events.

I found there is already an implementation of converting events to typed events. The client or the developer of sdk in another language may refer to this code:

https://github.com/line/lbm-sdk/blob/5eee50ecc2c08acbc02935b50d3b69e8ce3654eb/types/events.go#L104-L139

(It's not obvious whether it would be easy to port this code into another language, because it uses some packages)

With that function provided, one could easily restore the events to the corresponding typed events, like:

https://github.com/line/lbm-sdk/blob/5eee50ecc2c08acbc02935b50d3b69e8ce3654eb/types/events_test.go#L74-L103

With that aside, it will be a matter of taste to change the type of the message `EventModify`. If we decide to change the field to the singular one, I think we may also want to change the type of the field `changes` in the tx msg `MsgModify` to the singluar one too (which involves introducing a new tx endpoint and the corresponding msg).